# Import libraries

In [1]:
import os
import pandas as pd
import numpy as np

In [2]:
from pybaseball import pitching_stats

In [3]:
# Batting Stats , qual is the number of players appearences 
pitching = pitching_stats(2007,2022, qual=200)
pitching

IDfg  Season              Name Team  Age   W   L  WAR   ERA   G  ...  \
1    10954    2018      Jacob deGrom  NYM   30  10   9  9.0  1.70  32  ...   
12    1303    2011      Roy Halladay  PHI   34  19   6  8.7  2.35  32  ...   
6     1943    2009      Zack Greinke  KCR   25  16   8  8.7  2.16  33  ...   
4     2036    2015   Clayton Kershaw  LAD   27  16   7  8.6  2.13  33  ...   
203   8700    2009  Justin Verlander  DET   26  19   9  8.4  3.45  35  ...   
..     ...     ...               ...  ...  ...  ..  ..  ...   ...  ..  ...   
327   2072    2013    Jeremy Guthrie  KCR   34  15  12  0.6  4.04  33  ...   
321   3551    2012   Clayton Richard  SDP   28  14  14  0.4  3.99  33  ...   
258   4366    2011      Joe Saunders  ARI   30  12  13  0.3  3.69  33  ...   
384   1116    2007   Livan Hernandez  ARI   32  11  11  0.2  4.93  33  ...   
386    150    2010     Rodrigo Lopez  ARI   34   7  16 -0.1  5.00  33  ...   

       LA  Barrels  Barrel%  maxEV  HardHit  HardHit%  Events  CStr%   CSW%  \
1    11.1       20    0.039  112.9      148     0.287     515  0.164  0.315   
12    NaN        0      NaN    NaN        0       NaN       0  0.185  0.294   
6     NaN        0      NaN    NaN        0       NaN       0  0.187  0.287   
4     8.9       15    0.028  113.3      132     0.244     542  0.164  0.323   
203   NaN        0      NaN    NaN        0       NaN       0  0.175  0.289   
..    ...      ...      ...    ...      ...       ...     ...    ...    ...   
327   NaN        0      NaN    NaN        0       NaN       0  0.185  0.236   
321   NaN        0      NaN    NaN        0       NaN       0  0.183  0.246   
258   NaN        0      NaN    NaN        0       NaN       0  0.169  0.231   
384   NaN        0      NaN    NaN        0       NaN       0  0.185  0.239   
386   NaN        0      NaN    NaN        0       NaN       0  0.199  0.250   

     xERA  
1     NaN  
12    NaN  
6     NaN  
4     NaN  
203   NaN  
..    ...  
327   NaN  
321   NaN  
258   NaN  
384   NaN  
386   NaN  

[391 rows x 334 columns]

In [4]:
# Export DataFrame to csv
pitching.to_csv('../Data/Raw_Data/pitching.csv', index=False)

In [5]:
# Group by player ID and drop players with only 1 season of data 
# Leave players with a t least 2 seasons of data
pitching = pitching.groupby("IDfg", group_keys=False).filter(lambda x: x.shape[0]>1)

In [6]:
pitching

IDfg  Season              Name Team  Age   W   L  WAR   ERA   G  ...  \
1    10954    2018      Jacob deGrom  NYM   30  10   9  9.0  1.70  32  ...   
12    1303    2011      Roy Halladay  PHI   34  19   6  8.7  2.35  32  ...   
6     1943    2009      Zack Greinke  KCR   25  16   8  8.7  2.16  33  ...   
4     2036    2015   Clayton Kershaw  LAD   27  16   7  8.6  2.13  33  ...   
203   8700    2009  Justin Verlander  DET   26  19   9  8.4  3.45  35  ...   
..     ...     ...               ...  ...  ...  ..  ..  ...   ...  ..  ...   
343    976    2010        Randy Wolf  MIL   33  13  12  0.8  4.17  34  ...   
327   2072    2013    Jeremy Guthrie  KCR   34  15  12  0.6  4.04  33  ...   
321   3551    2012   Clayton Richard  SDP   28  14  14  0.4  3.99  33  ...   
258   4366    2011      Joe Saunders  ARI   30  12  13  0.3  3.69  33  ...   
384   1116    2007   Livan Hernandez  ARI   32  11  11  0.2  4.93  33  ...   

       LA  Barrels  Barrel%  maxEV  HardHit  HardHit%  Events  CStr%   CSW%  \
1    11.1       20    0.039  112.9      148     0.287     515  0.164  0.315   
12    NaN        0      NaN    NaN        0       NaN       0  0.185  0.294   
6     NaN        0      NaN    NaN        0       NaN       0  0.187  0.287   
4     8.9       15    0.028  113.3      132     0.244     542  0.164  0.323   
203   NaN        0      NaN    NaN        0       NaN       0  0.175  0.289   
..    ...      ...      ...    ...      ...       ...     ...    ...    ...   
343   NaN        0      NaN    NaN        0       NaN       0  0.181  0.248   
327   NaN        0      NaN    NaN        0       NaN       0  0.185  0.236   
321   NaN        0      NaN    NaN        0       NaN       0  0.183  0.246   
258   NaN        0      NaN    NaN        0       NaN       0  0.169  0.231   
384   NaN        0      NaN    NaN        0       NaN       0  0.185  0.239   

     xERA  
1     NaN  
12    NaN  
6     NaN  
4     NaN  
203   NaN  
..    ...  
343   NaN  
327   NaN  
321   NaN  
258   NaN  
384   NaN  

[326 rows x 334 columns]

In [7]:
# Write a function that takes WAR number from a player's season to insert in Next_WAR column
def next_season(player):
    player = player.sort_values("Season")
    player["Next_WAR"] = player["WAR"].shift(-1)
    return player

pitching = pitching.groupby("IDfg", group_keys=False).apply(next_season)

In [8]:
# Show the next WAR of each player in a DF 
# Missing values are for some seasons with no data (some players could not play in these seasons)
war_df=pitching[['Name','Season','WAR','Next_WAR']]
war_df

Name  Season  WAR  Next_WAR
241     Mark Buehrle    2007  3.3       4.2
286     Mark Buehrle    2008  4.2       3.0
295     Mark Buehrle    2009  3.0       3.8
353     Mark Buehrle    2010  3.8       3.1
237     Mark Buehrle    2011  3.1       1.6
..               ...     ...  ...       ...
160       Aaron Nola    2022  6.3       NaN
139  Sandy Alcantara    2021  4.3       5.7
11   Sandy Alcantara    2022  5.7       NaN
164     Shane Bieber    2019  5.5       4.9
78      Shane Bieber    2022  4.9       NaN

[326 rows x 4 columns]

## Cleaning Data

In [9]:
# Check null values
null_data=pitching.isnull().sum()
null_data

IDfg          0
Season        0
Name          0
Team          0
Age           0
           ... 
Events        0
CStr%         0
CSW%          0
xERA        326
Next_WAR     91
Length: 335, dtype: int64

In [10]:
# List columns with no missing values 
complete_cols=list(pitching.columns[null_data == 0])

In [11]:
# Create a clean DataFrame with complete columns and Next_WAR column
pitching = pitching[complete_cols + ["Next_WAR"]].copy()
pitching

IDfg  Season             Name Team  Age   W   L  WAR   ERA   G  ...  \
241    225    2007     Mark Buehrle  CHW   28  10   9  3.3  3.63  30  ...   
286    225    2008     Mark Buehrle  CHW   29  15  12  4.2  3.79  34  ...   
295    225    2009     Mark Buehrle  CHW   30  13  10  3.0  3.84  33  ...   
353    225    2010     Mark Buehrle  CHW   31  13  13  3.8  4.28  33  ...   
237    225    2011     Mark Buehrle  CHW   32  13   9  3.1  3.59  31  ...   
..     ...     ...              ...  ...  ...  ..  ..  ...   ...  ..  ...   
160  16149    2022       Aaron Nola  PHI   29  11  13  6.3  3.25  32  ...   
139  18684    2021  Sandy Alcantara  MIA   25   9  15  4.3  3.19  33  ...   
11   18684    2022  Sandy Alcantara  MIA   26  14   9  5.7  2.28  32  ...   
164  19427    2019     Shane Bieber  CLE   24  15   8  5.5  3.28  34  ...   
78   19427    2022     Shane Bieber  CLE   27  13   8  4.9  2.88  31  ...   

     Oppo%+  Soft%+  Med%+  Hard%+  Barrels  HardHit  Events  CStr%   CSW%  \
241      81     111    100      94        0        0       0  0.171  0.250   
286      83     115    100      91        0        0       0  0.171  0.239   
295      95      98    109      81        0        0       0  0.176  0.242   
353     100     124     99      87        0        0       0  0.181  0.242   
237      85     115     95      97        0        0       0  0.179  0.244   
..      ...     ...    ...     ...      ...      ...     ...    ...    ...   
160     104     103    100      98       38      169     534  0.197  0.324   
139      91     116    100      91       35      227     576  0.146  0.279   
11       93     126     99      87       32      239     620  0.153  0.275   
164     104      79     97     112       41      231     554  0.185  0.325   
78       92      74    102     111       40      238     555  0.175  0.314   

     Next_WAR  
241       4.2  
286       3.0  
295       3.8  
353       3.1  
237       1.6  
..        ...  
160       NaN  
139       5.7  
11        NaN  
164       4.9  
78        NaN  

[326 rows x 155 columns]

### Drop and Dummify all of the categorical variables 

In [12]:
#Check columns types 
pitching.dtypes

IDfg          int64
Season        int64
Name         object
Team         object
Age           int64
             ...   
HardHit       int64
Events        int64
CStr%       float64
CSW%        float64
Next_WAR    float64
Length: 155, dtype: object

In [13]:
# Find columns with object data type
pitching.dtypes[pitching.dtypes == 'object']

Name       object
Team       object
Dollars    object
Age Rng    object
dtype: object

In [15]:
# Drop 'Dollars' (Dollar value of player) and 'Age Rng' (Player's age range during a season) 
pitching = pitching.drop(['Dollars','Age Rng'], axis = 1, inplace = False)

In [16]:
# Assign each team name to number 
pitching["Team_code"] = pitching["Team"].astype("category").cat.codes

In [17]:
# Copy pitching data 
pitching_full = pitching.copy()

# Drop Nan Values from 'Next_Raw'
pitching = pitching.dropna().copy()

In [18]:
pitching

IDfg  Season             Name Team  Age   W   L  WAR   ERA   G  ...  \
241    225    2007     Mark Buehrle  CHW   28  10   9  3.3  3.63  30  ...   
286    225    2008     Mark Buehrle  CHW   29  15  12  4.2  3.79  34  ...   
295    225    2009     Mark Buehrle  CHW   30  13  10  3.0  3.84  33  ...   
353    225    2010     Mark Buehrle  CHW   31  13  13  3.8  4.28  33  ...   
237    225    2011     Mark Buehrle  CHW   32  13   9  3.1  3.59  31  ...   
..     ...     ...              ...  ...  ...  ..  ..  ...   ...  ..  ...   
360  13431    2016   Marcus Stroman  TOR   25   9  10  3.3  4.37  32  ...   
13   16149    2018       Aaron Nola  PHI   25  17   6  5.5  2.37  33  ...   
301  16149    2019       Aaron Nola  PHI   26  12   7  3.4  3.87  34  ...   
139  18684    2021  Sandy Alcantara  MIA   25   9  15  4.3  3.19  33  ...   
164  19427    2019     Shane Bieber  CLE   24  15   8  5.5  3.28  34  ...   

     Soft%+  Med%+  Hard%+  Barrels  HardHit  Events  CStr%   CSW%  Next_WAR  \
241     111    100      94        0        0       0  0.171  0.250       4.2   
286     115    100      91        0        0       0  0.171  0.239       3.0   
295      98    109      81        0        0       0  0.176  0.242       3.8   
353     124     99      87        0        0       0  0.181  0.242       3.1   
237     115     95      97        0        0       0  0.179  0.244       1.6   
..      ...    ...     ...      ...      ...     ...    ...    ...       ...   
360      98    100     101       28      264     631  0.172  0.264       3.4   
13      122    113      71       21      168     542  0.198  0.322       3.4   
301      92     95     110       34      210     532  0.206  0.315       6.3   
139     116    100      91       35      227     576  0.146  0.279       5.7   
164      79     97     112       41      231     554  0.185  0.325       4.9   

     Team_code  
241          6  
286          6  
295          6  
353          6  
237          6  
..         ...  
360         31  
13          22  
301         22  
139         16  
164          8  

[235 rows x 154 columns]

## Machine Learning Model to predict next WAR

In [26]:
from sklearn.linear_model import Ridge
from sklearn.feature_selection import SequentialFeatureSelector
from sklearn.model_selection import TimeSeriesSplit

# Initialize Ridge Regression Model
# If Lambda or alpa in Python is higher it reduces overfitting, if it's lower it's closer to regular linear regression
rr = Ridge(alpha=1)


# Split Data into 3 parts 
split = TimeSeriesSplit(n_splits=3)

# Go over the features to find the best one until it selects 20 
sfs = SequentialFeatureSelector(rr, 
                                n_features_to_select=20, 
                                direction="forward",
                                cv=split,
                                n_jobs=8
                               )

In [27]:
# Remove some columns 
removed_columns = ["Next_WAR", "Name", "Team", "IDfg", "Season"]
selected_columns = pitching.columns[~pitching.columns.isin(removed_columns)]

In [28]:
# All columns except the one we have removed 
selected_columns

Index(['Age', 'W', 'L', 'WAR', 'ERA', 'G', 'GS', 'CG', 'ShO', 'SV',
       ...
       'Oppo%+', 'Soft%+', 'Med%+', 'Hard%+', 'Barrels', 'HardHit', 'Events',
       'CStr%', 'CSW%', 'Team_code'],
      dtype='object', length=149)

In [29]:
X=pitching[selected_columns]
y=pitching['Next_WAR']

In [30]:
# Scale the data 
# MinMax Scales values to be between 0 and 1
from sklearn.preprocessing import MinMaxScaler

scaler = MinMaxScaler()
pitching.loc[:,selected_columns] = scaler.fit_transform(X)

In [31]:
# Most values are now between 0 and 1
pitching.describe()

IDfg       Season         Age           W           L  \
count    235.000000   235.000000  235.000000  235.000000  235.000000   
mean    4383.293617  2011.459574    0.402128    0.472813    0.423404   
std     3694.105282     3.121723    0.190267    0.189575    0.183224   
min      225.000000  2007.000000    0.000000    0.000000    0.000000   
25%     1571.500000  2009.000000    0.277778    0.333333    0.312500   
50%     3254.000000  2011.000000    0.388889    0.500000    0.437500   
75%     6345.000000  2014.000000    0.555556    0.611111    0.562500   
max    19427.000000  2021.000000    1.000000    1.000000    1.000000   

              WAR         ERA           G          GS          CG  ...  \
count  235.000000  235.000000  235.000000  235.000000  235.000000  ...   
mean     0.466296    0.476003    0.594681    0.668085    0.188008  ...   
std      0.182282    0.177600    0.145530    0.162674    0.180151  ...   
min      0.000000    0.000000    0.000000    0.000000    0.000000  ...   
25%      0.352273    0.362216    0.500000    0.571429    0.090909  ...   
50%      0.465909    0.471591    0.625000    0.714286    0.090909  ...   
75%      0.579545    0.589489    0.750000    0.714286    0.272727  ...   
max      1.000000    1.000000    1.000000    1.000000    1.000000  ...   

           Soft%+       Med%+      Hard%+     Barrels     HardHit      Events  \
count  235.000000  235.000000  235.000000  235.000000  235.000000  235.000000   
mean     0.529730    0.539116    0.502837    0.116538    0.123694    0.144216   
std      0.175608    0.178639    0.172765    0.261058    0.273323    0.316218   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      0.391892    0.423077    0.398148    0.000000    0.000000    0.000000   
50%      0.513514    0.538462    0.500000    0.000000    0.000000    0.000000   
75%      0.662162    0.653846    0.592593    0.000000    0.000000    0.000000   
max      1.000000    1.000000    1.000000    1.000000    1.000000    1.000000   

            CStr%        CSW%    Next_WAR   Team_code  
count  235.000000  235.000000  235.000000  235.000000  
mean     0.504945    0.476498    4.164681    0.468351  
std      0.184939    0.163777    1.804975    0.314981  
min      0.000000    0.000000    0.300000    0.000000  
25%      0.378378    0.369231    2.800000    0.187500  
50%      0.486486    0.476923    4.100000    0.437500  
75%      0.621622    0.561538    5.400000    0.781250  
max      1.000000    1.000000    9.000000    1.000000  

[8 rows x 152 columns]

In [32]:
# Fit the data 
sfs.fit(X, y)

SequentialFeatureSelector(cv=TimeSeriesSplit(gap=0, max_train_size=None, n_splits=3, test_size=None),
                          estimator=Ridge(alpha=1), n_features_to_select=20,
                          n_jobs=8)

In [33]:
# sfs.get_support() returns True to each column selected 
sfs.get_support()

array([False, False,  True,  True, False,  True, False, False, False,
        True,  True, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
        True, False, False, False, False,  True, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False, False, False, False, False, False,  True,  True,
        True,  True, False, False, False,  True, False, False, False,
       False, False,  True, False, False, False, False, False, False,
       False, False, False, False, False, False, False,  True, False,
       False,  True, False,  True, False, False, False, False, False,
       False,  True, False,  True, False, False, False, False, False,
       False, False, False, False, False, False, False, False, False,
       False, False,

In [36]:
# Assign sfs to a variable called predictors
# These are the 20 columns selected by our model
predictors = list(selected_columns[sfs.get_support()])
predictors

['L',
 'WAR',
 'G',
 'SV',
 'BS',
 'LD%',
 'IFH%',
 'FBv',
 'wFB',
 'wFB/C',
 'O-Swing%',
 'Z-Contact%',
 'SD',
 'Z-Contact% (sc)',
 'Pace',
 'BIP-Wins',
 'Soft%',
 'Hard%',
 'Cent%+',
 'Soft%+']

In [37]:
# We only want to use past data to predict future data 

def backtest(data, model, predictors, start=5, step=1):
    # Create a list to hold all seasons predictions
    all_predictions = []
    
    # List of seasons sorted by order 
    years = sorted(data["Season"].unique())
    
    for i in range(start, len(years), step):
        current_year = years[i]
        
        # Train set is all years before current_year
        train = data[data["Season"] < current_year]
        
        # Test set is the current_year
        test = data[data["Season"] == current_year]
        
        model.fit(train[predictors], train["Next_WAR"])
        
        preds = model.predict(test[predictors])
        preds = pd.Series(preds, index=test.index)
        combined = pd.concat([test["Next_WAR"], preds], axis=1)
        combined.columns = ["actual", "prediction"]
        
        all_predictions.append(combined)
    return pd.concat(all_predictions)

In [38]:
model_pred=backtest(pitching, rr, predictors)
model_pred

actual  prediction
271     2.2    2.093647
192     2.1    4.617431
222     1.7    3.592745
274     1.2    2.493058
190     1.5    4.338214
..      ...         ...
13      3.4    4.971226
30      3.3    7.581381
301     6.3    3.924974
164     4.9    4.905385
139     5.7    5.334379

[106 rows x 2 columns]

## Caculate the accuracy

In [39]:
# Calculate the mean squared error value (To compare the prediction from actual value)
from sklearn.metrics import mean_squared_error

mean_squared_error(model_pred["actual"], model_pred["prediction"])

2.0647282578808364

In [40]:
# Square root of mean_squared_error
2.6490416143593762 ** 0.5

1.6275876671809038

In [41]:
pitching['Next_WAR'].describe()

count    235.000000
mean       4.164681
std        1.804975
min        0.300000
25%        2.800000
50%        4.100000
75%        5.400000
max        9.000000
Name: Next_WAR, dtype: float64

std = 1.804975

Square root of mean_squared_error = 1.6275876671809038 < std 
Which is good for the model

### Add player history data to improve accuracy

In [51]:
def player_history(df):
    df = df.sort_values("Season")
        
    df["player_season"] = range(0, df.shape[0])
    # Find the corr between "player_season" and "WAR"
    df["war_corr"] = list(df[["player_season", "WAR"]].expanding().corr().loc[(slice(None), "player_season"),"WAR"])
    df["war_corr"].fillna(0, inplace=True)
    
    df["war_diff"] = df["WAR"] / df["WAR"].shift(1)
    # Replace NaN values with 1 when there's no previous season 
    df["war_diff"].fillna(1, inplace=True)
    # Replace infinite value with 1
    df["war_diff"][df["war_diff"] == np.inf] = 1
    
    return df

pitching = pitching.groupby("IDfg", group_keys=False).apply(player_history)


In [52]:
# Find averages across seasos; tell us if a player performed better than the average of a season or no 
def group_averages(df):
    return df["WAR"] / df["WAR"].mean()

In [53]:
pitching["war_season"] = pitching.groupby("Season", group_keys=False).apply(group_averages)

In [54]:
new_predictors = predictors + ["player_season", "war_corr", "war_season", "war_diff"]

In [55]:
predictions = backtest(pitching, rr, new_predictors)

In [56]:
# Calculate the mean squared error value (To compare the prediction from actual value)
mean_squared_error(predictions["actual"], predictions["prediction"]) 

2.188700421177309

Value slighly bigger than before

In [57]:
pd.Series(rr.coef_, index=new_predictors).sort_values()

G                 -1.051227
BIP-Wins          -1.039589
IFH%              -0.939685
Z-Contact%        -0.687934
Cent%+            -0.671681
Z-Contact% (sc)   -0.634823
O-Swing%          -0.426066
LD%               -0.296720
Soft%             -0.291252
war_corr          -0.172216
war_diff          -0.129758
BS                 0.000000
SV                 0.000000
player_season      0.013012
wFB                0.334417
Pace               0.416059
wFB/C              0.618330
L                  0.721273
Soft%+             0.808778
Hard%              0.898425
WAR                1.267573
SD                 1.378813
war_season         1.695461
FBv                1.852224
dtype: float64

All small values indicated that the model is not taking these columns into account to make the prediction.

In [60]:
merged = predictions.merge(pitching, left_index=True, right_index=True)
merged["diff"] = (predictions["actual"] - predictions["prediction"]).abs()
next_war_df=merged[["IDfg", "Season", "Name", "WAR", "Next_WAR", "diff"]].sort_values(["diff"])
next_war_df.head(20)

IDfg  Season               Name       WAR  Next_WAR      diff
84    3137    2013       Max Scherzer  0.647727       5.6  0.029588
263   1245    2014        R.A. Dickey  0.181818       1.8  0.051278
113   4772    2012    Felix Hernandez  0.727273       6.1  0.062359
14    2233    2014    Adam Wainwright  0.534091       3.8  0.066656
211   1943    2012       Zack Greinke  0.522727       4.5  0.070315
7     6893    2014       Johnny Cueto  0.477273       4.1  0.086056
238   4972    2013        Cole Hamels  0.511364       4.5  0.088866
8     2429    2017       Corey Kluber  0.795455       5.5  0.089496
24    4930    2014         Jon Lester  0.590909       4.9  0.099183
89    5524    2015  Madison Bumgarner  0.534091       4.3  0.099336
175  11423    2014      Jose Quintana  0.534091       4.3  0.129786
107   4772    2013    Felix Hernandez  0.670455       5.6  0.134811
1    10954    2018       Jacob deGrom  1.000000       6.9  0.138915
22    2429    2014       Corey Kluber  0.795455       5.6  0.150603
360  13431    2016     Marcus Stroman  0.352273       3.4  0.151281
214   2429    2015       Corey Kluber  0.613636       4.9  0.191723
156   3200    2013      Ervin Santana  0.261364       2.7  0.200515
64    3137    2015       Max Scherzer  0.715909       5.6  0.209231
317   2520    2013         Lance Lynn  0.352273       3.6  0.211686
93    2233    2013    Adam Wainwright  0.681818       4.9  0.216852

In [61]:
next_war_df = next_war_df.groupby("IDfg", group_keys=False).apply(next_season)
next_war_df

IDfg  Season             Name       WAR  Next_WAR      diff
271    225    2012     Mark Buehrle  0.159091  0.227273  0.364429
340    225    2013     Mark Buehrle  0.227273       NaN  1.886665
192    404    2012      CC Sabathia  0.443182       NaN  3.107495
222    512    2012     A.J. Burnett  0.375000       NaN  2.112238
274    978    2012   Bronson Arroyo  0.227273       NaN  1.381573
..     ...     ...              ...       ...       ...       ...
360  13431    2016   Marcus Stroman  0.352273       NaN  0.151281
13   16149    2018       Aaron Nola  0.602273  0.363636  1.081038
301  16149    2019       Aaron Nola  0.363636       NaN  2.547625
139  18684    2021  Sandy Alcantara  0.465909       NaN  0.473284
164  19427    2019     Shane Bieber  0.602273       NaN  0.372380

[106 rows x 6 columns]

In [ ]:
import json
next_war_js=next_war_df.to_json(orient = "records")
next_war_js

In [ ]:
# Save JSON file in clean data file
save_file=open('../Data/Clean_Data/next_war.js','w')
json.dump(next_war_js, save_file, indent = 6)  
save_file.close() 

# Pitching Stats & Records Data

In [ ]:
# Add an option on website for user to check the records of previous games 
data_test = schedule_and_record(2017, 'NYY')
data_test.head()